### PR TITLE
Cans-819 Implement Front-End Assessment Delete

### PR DESCRIPTION
**Pull request by ggomez-cambriasolutions**
## JIRA
https://osi-cwds.atlassian.net/browse/CANS-819

## Description
Front end changes for assessment Delete functionality. Call the new endpoint, implement new logic, warning modal, & assessment card status changes.

## Tests

- [x] I used TDD to develop the feature
- [x] I have included unit tests
- [ ] I have included new acceptance tests or modified existing ones
- [ ] I have included other tests

## Types of changes

- [x] New feature (non-breaking change which adds functionality)
- [x] Refactoring (No behavioral changes)
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Checklist:

- [x] My code follows the code style of this project.
- [x] I ran all my tests locally which were green.
- [x] My code adds no new issues to Code Climate
- [x] I ran all the linters for the project.
- [x] My code is in a stable state ready to be deployable, but not necessarily complete.
- [x] I promise on my honor as a CWDS developer to ensure I don't break the build (or bring donuts if I leave things broken), to check linters, use best practices, to leave the code in better shape than I found it, and to have a good day.

### CODE REVIEW

File: app/javascript/Application/components/common/Ellipsis.js [hunk 1]
```diff
@@ -23,14 +28,103 @@ export default class Ellipsis extends React.Component {
     })
   }
 
+  onItemSelect(e) {
```
Comment (pokharelmanish) [inline app/javascript/Application/components/common/Ellipsis.js]: we can use arrow function here so we don't need to bind in the constructor.

File: app/javascript/Application/components/common/Ellipsis.test.js [hunk 2]
```diff
@@ -2,65 +2,244 @@ import React from 'react'
...
+      })
     })
 
-    describe('Link component', () => {
-      it('renders a Link component and the correct props', () => {
-        const wrapper = shallow(<Ellipsis {...defaultProps} />)
-        expect(wrapper.find(Link).props().to).toEqual({
-          pathname: '/staff/0X5/clients/C76Jg230X3/assessments/1234/changelog',
+    describe('PopOver Menu Items', () => {
+      let popOverBody
```
Comment (pokharelmanish) [inline app/javascript/Application/components/common/Ellipsis.test.js]: `popoverBody`?

File: app/controllers/api/assessments_controller.rb [hunk 3]
```diff
@@ -27,5 +27,10 @@ def changes
       response = Assessments::AssessmentsRepository.new(session[:token]).changes(params[:id])
       render json: response.body, status: response.status
     end
+
+    def delete
```
Comment (pokharelmanish) [inline app/controllers/api/assessments_controller.rb]: Do we have a test for this?

File: app/javascript/Application/components/Assessment/ChangeLog/ChangeLogLoadingBoundary.js [hunk 4]
```diff
@@ -1,19 +1,10 @@
...
-  return Promise.all([changeHistoryPromise, assessmentPromise]).then(([changeHistory, assessment]) => ({
-    changeHistory,
-    assessment,
-  }))
-}
+import AssessmentService from '../Assessment.service'
 
 const ChangeLogLoadingBoundary = ({ id, children }) => (
-  <LoadingBoundary childNodeFetchedPropName={'assessmentWithHistory'} fetch={() => fetchBoth(id)}>
+  <LoadingBoundary childNodeFetchedPropName={'assessmentHistory'} fetch={() => AssessmentService.getAllChanges(id)}>
```
Comment (wmitchell-cambria) [inline app/javascript/Application/components/Assessment/ChangeLog/ChangeLogLoadingBoundary.js]: Re: the merge conflicts in these LoadingBoundaries, my search refactors added the ability for LoadingBoundaries to update with new information. They do this when their fetch prop changes. So, passing an anonymous function like this doesn't work as intended any more, since it will create a new function (and therefore fetch new data) every single render.

In most cases, I created a 'fetchFactory' that saves a fetch in state. To solve the conflict, we just need to make the same changes to that fetchFactory that you have here.

File: app/javascript/Application/components/Client/AssessmentHistory/ClientAssessmentHistory.js [hunk 5]
```diff
@@ -1,31 +1,37 @@
...
+    const { callback } = this.props
+    const loadingState = this.props.loadingState ? this.props.loadingState : ''
+    const noDataText =
+      loadingState === LoadingState.ready
+        ? 'No assessments currently exist for this child/youth.'
+        : 'Loading assessments...'
 
     return assessments.length === 0 ? (
-      <div id="no-data">No assessments currently exist for this child/youth.</div>
+      <div id="no-data">{noDataText}</div>
```
Comment (pokharelmanish) [inline app/javascript/Application/components/Client/AssessmentHistory/ClientAssessmentHistory.js]: probably we need padding here. 
`padding-left: 1.5rem`

**Comment by pokharelmanish** [comment]
retest this please

File: app/javascript/Application/components/Assessment/ChangeLog/AssessmentChangeLog.js [hunk 6]
```diff
@@ -38,51 +38,47 @@ class AssessmentChangeLog extends Component {
...
     pageHeaderButtonsController.updateHeaderButtons(leftButton, rightButton)
   }
 
-  updatePrintButtonIfNeeded(prevChangeHistory) {
-    const { assessmentWithHistory } = this.props
-    const changeHistory = assessmentWithHistory.changeHistory ? assessmentWithHistory.changeHistory : []
-    const changeHistoryUpdated = prevChangeHistory !== changeHistory
+  updatePrintButtonIfNeeded(prevAssessmentHistory) {
+    const { assessmentHistory } = this.props
+    const assessmentHistoryUpdated = Boolean(prevAssessmentHistory !== assessmentHistory)
```
Comment (wmitchell-cambria) [inline app/javascript/Application/components/Assessment/ChangeLog/AssessmentChangeLog.js]: nit: This seems a little verbose to me. `!==` will already be a Boolean, so the cast is unnecessary. (Besides, even if it were not a Boolean, it is only used in the conditional anyways...) Why not just inline it all into the `if` statement?

File: app/javascript/Application/components/Assessment/ChangeLog/AssessmentChangeLog.js [hunk 7]
```diff
@@ -38,51 +38,47 @@ class AssessmentChangeLog extends Component {
...
   }
 
   shouldPrintButtonBeEnabled() {
-    const { assessmentWithHistory } = this.props
-    const changeHistory = assessmentWithHistory.changeHistory ? assessmentWithHistory.changeHistory : []
-    const changeHistoryLength = changeHistory.length
-    const enablePrintButton = changeHistoryLength > 0
+    const { assessmentHistory } = this.props
+    const assessmentHistoryLength = assessmentHistory.length
+    const enablePrintButton = assessmentHistoryLength > 0
```
Comment (wmitchell-cambria) [inline app/javascript/Application/components/Assessment/ChangeLog/AssessmentChangeLog.js]: nit: Just return it here instead of setting it to a variable?

I think some of this code is coming out like this because it is simplifying what used to be more complex. Sometimes it helps to step back and look at the final result with fresh eyes.

File: app/javascript/Application/components/Assessment/ChangeLog/AssessmentChangeLog.js [hunk 8]
```diff
@@ -132,10 +125,7 @@ class AssessmentChangeLog extends Component {
 }
 
 AssessmentChangeLog.propTypes = {
-  assessmentWithHistory: PropTypes.shape({
-    changeHistory: PropTypes.arrayOf(changeHistoryPropTypes),
-    assessment: PropTypes.object,
-  }),
+  assessmentHistory: PropTypes.arrayOf(assessmentHistoryPropTypes),
```
Comment (wmitchell-cambria) [inline app/javascript/Application/components/Assessment/ChangeLog/AssessmentChangeLog.js]: I hadn't realized how much extra logic there was because of this complex prop. It's so clean after these changes!

File: app/javascript/Application/components/Assessment/assessment.mocks.test.js [hunk 9]
```diff
@@ -980,6 +1039,37 @@ export const assessmentCompletedWithReferralNumber = {
...
+  updated_by: {
+    first_name: 'Name 2',
+    last_name: 'LastName 2',
+  },
+  county: {
+    name: 'Alameda',
+  },
+}
+
+export const assessmentDeletedWithReferralNumber = {
```
Comment (wmitchell-cambria) [inline app/javascript/Application/components/Assessment/assessment.mocks.test.js]: I'm starting to think we need a better way of defining these... Maybe we should make a tech debt story? Having thousands of lines of mock data like this can be tedious to manage, as a change in a particular field needs to be applied to each one. It also hides a lot of details from the tests themselves, and those details may or may not be important.

It's possible that something like FactoryBot would be helpful here, where we could construct mock assessments using descriptive code. `assessment.withStatusDeleted().withReferralNumber()`

File: app/javascript/Application/components/Client/AssessmentHistory/ClientAssessmentHistory.js [hunk 10]
```diff
@@ -1,31 +1,37 @@
...
         <AddCansLink clientIdentifier={clientIdentifier} />
       </AuthBoundary>
-    )
+    ) : null
   }
 
   renderAssessments = assessments => {
     const assessmentsToDisplay = 3
+    const { callback } = this.props
+    const loadingState = this.props.loadingState ? this.props.loadingState : ''
```
Comment (wmitchell-cambria) [inline app/javascript/Application/components/Client/AssessmentHistory/ClientAssessmentHistory.js]: nit: Would something bad happen if `this.props.loadingState` were not defined? Why do we default it to an empty string?

File: app/javascript/Application/components/Client/AssessmentHistory/ClientAssessmentHistory.js [hunk 11]
```diff
@@ -84,8 +93,10 @@ class ClientAssessmentHistory extends PureComponent {
 
 ClientAssessmentHistory.propTypes = {
   assessments: PropTypes.arrayOf(PropTypes.object),
+  callback: PropTypes.func.isRequired,
```
Comment (wmitchell-cambria) [inline app/javascript/Application/components/Client/AssessmentHistory/ClientAssessmentHistory.js]: nit: Callback for what? It's required, but there is not much info about when it is called or why, which could help the parent author to decide what function is appropriate to set.

File: app/javascript/Application/components/Client/AssessmentHistory/ClientAssessmentHistory.test.js [hunk 12]
```diff
@@ -131,5 +135,16 @@ describe('<ClientAssessmentHistory', () => {
...
     })
+
+    describe('when assessments are loading', () => {
+      it('renders the loading message', () => {
+        // given + when
+        const wrapper = getWrapper([], LoadingState.waiting)
+
+        // then
+        const message = wrapper.find('#no-data').text()
+        expect(message).toBe('Loading assessments...')
```
Comment (wmitchell-cambria) [inline app/javascript/Application/components/Client/AssessmentHistory/ClientAssessmentHistory.test.js]: Is this related to the Delete story? It seems like handling this loading state roped in a couple hundred lines of code changes that could have been split out. It's harder as a reviewer to focus on a PR that covers multiple topics.

File: app/javascript/Application/components/Client/AssessmentHistory/ClientAssessmentHistoryLoadingBoundary.js [hunk 13]
```diff
@@ -4,9 +4,14 @@ import LoadingBoundary from '../../common/LoadingBoundary'
 import { AssessmentService } from '../../Assessment/'
 
 const ClientAssessmentHistoryLoadingBoundary = ({ clientIdentifier, children }) => {
-  const service = () => AssessmentService.search({ client_identifier: clientIdentifier })
+  const service = () =>
```
Comment (wmitchell-cambria) [inline app/javascript/Application/components/Client/AssessmentHistory/ClientAssessmentHistoryLoadingBoundary.js]: Same comment as before. This should not be created each and every render. Only when the identifier changes.

File: app/javascript/Application/components/Client/AssessmentHistory/ClientAssessmentHistoryTable.js [hunk 14]
```diff
@@ -9,21 +9,28 @@ import ClientAssessmentHistoryTableDate from './ClientAssessmentHistoryTableDate
 import ClientAssessmentHistoryTableUpdatedBy from './ClientAssessmentHistoryTableUpdatedBy'
 import ClientAssessmentHistoryTableEllipsis from './ClientAssessmentHistoryTableEllipsis'
 
-const columnWidths = [190, 250, 140, 170, 260, 35]
+const columnWidths = [0, 190, 250, 140, 170, 260, 35]
```
Comment (wmitchell-cambria) [inline app/javascript/Application/components/Client/AssessmentHistory/ClientAssessmentHistoryTable.js]: I can't say for sure that rendering a 0 width comment just as a backup sorting option is a bad idea... but it _does_ seem odd to me. Are we sure we need to do this? How often are there multiple assessments on the same day, and could the Updated Time be a suitable replacement for those rare cases?

File: app/javascript/Application/components/Client/AssessmentHistory/ClientAssessmentHistoryTableEllipsis.js [hunk 15]
```diff
@@ -3,14 +3,31 @@ import PropTypes from 'prop-types'
 import Ellipsis from '../../common/Ellipsis'
 
 const ClientAssessmentHistoryTableEllipsis = ({ original }) => {
-  const { person, id, inheritUrl } = original
-  return id && person ? <Ellipsis clientId={person.identifier} inheritUrl={inheritUrl} assessmentId={id} /> : null
+  const { person, id, metadata, status, inheritUrl, callback } = original
+  const shouldRender = Boolean(person && id && metadata && status && inheritUrl && callback)
```
Comment (wmitchell-cambria) [inline app/javascript/Application/components/Client/AssessmentHistory/ClientAssessmentHistoryTableEllipsis.js]: Is there a valid situation in which these properties may be falsy?

File: app/controllers/api/assessments_controller.rb (hunk 3 above)
Comment (ggomez-cambriasolutions) [reply app/controllers/api/assessments_controller.rb]: yes

File: app/javascript/Application/components/Client/AssessmentHistory/ClientAssessmentHistory.js (hunk 5 above)
Comment (ggomez-cambriasolutions) [reply app/javascript/Application/components/Client/AssessmentHistory/ClientAssessmentHistory.js]: we already define padding in the sass file, both here and the search assessment history.

File: app/javascript/Application/components/Assessment/ChangeLog/AssessmentChangeLog.js (hunk 7 above)
Comment (ggomez-cambriasolutions) [reply app/javascript/Application/components/Assessment/ChangeLog/AssessmentChangeLog.js]: agreed

File: app/javascript/Application/components/Assessment/ChangeLog/AssessmentChangeLog.js (hunk 8 above)
Comment (ggomez-cambriasolutions) [reply app/javascript/Application/components/Assessment/ChangeLog/AssessmentChangeLog.js]: right? Igor made some improvements to the API, so we didn't have to fetch the assessment and the assessment history with two API calls.

File: app/javascript/Application/components/Assessment/assessment.mocks.test.js (hunk 9 above)
Comment (ggomez-cambriasolutions) [reply app/javascript/Application/components/Assessment/assessment.mocks.test.js]: Yes, we definitely need to clean up this file. It has gotten way too big.

File: app/javascript/Application/components/Client/AssessmentHistory/ClientAssessmentHistory.js (hunk 10 above)
Comment (ggomez-cambriasolutions) [reply app/javascript/Application/components/Client/AssessmentHistory/ClientAssessmentHistory.js]: nothing bad would happen. I was just being over cautious, but it's fine just destructuring the loadingState from the props object.

File: app/javascript/Application/components/Client/AssessmentHistory/ClientAssessmentHistoryTable.js (hunk 14 above)
Comment (ggomez-cambriasolutions) [reply app/javascript/Application/components/Client/AssessmentHistory/ClientAssessmentHistoryTable.js]: I had to add the assessment id column, because the table without it, after deleting assessment, the assessment record would move to a different row within the same event date grouping, so I had to sort by event date, and then assessment id. That is the purpose of having the hidden column, because the design does not state to show that column. Using the updated by date wouldn't solve this problem of having the record show up in a different spot after deleting and fetching assessment history again. It's not a backup sorting option, it is always used.

Comment (wmitchell-cambria) [reply app/javascript/Application/components/Client/AssessmentHistory/ClientAssessmentHistoryTable.js]: Ah, interesting, that makes sense.  Would another option be to use Created timestamp as the secondary sort field, which wouldn't change when the assessment is deleted? (By 'backup', I just meant that it doesn't affect the sort unless the assessment dates are the same. It's a secondary field.)

File: app/javascript/Application/components/Search/SearchAssessmentHistory.js [hunk 16]
```diff
@@ -3,22 +3,30 @@ import PropTypes from 'prop-types'
...
+import { AssessmentStatus, sortAssessments } from '../Assessment'
 
 class SearchAssessmentHistory extends Component {
   constructor(props) {
     super(props)
-
     this.state = {
       assessments: [],
       fetchStatus: LoadingState.idle,
+      lastDeletedAssessmentId: 0,
```
[the remark this answers is not in the document]
Comment (ggomez-cambriasolutions) [reply app/javascript/Application/components/Search/SearchAssessmentHistory.js]: The assessment Id is an integer. I made a loading boundary for search assessment history, to help with updating the assessment history when we delete an assessment on the search page. So, we don't need lastDeletedAssessmentId in state anymore.

File: app/javascript/Application/styles/global/colors.sass [hunk 17]
```diff
@@ -31,6 +32,7 @@ $burgundy: #AD2213
 
 // Black
 $transparent-black: rgba(0, 0, 0, .7)
+$transparent-black-light: rgba(0, 0, 0, .2)
```
[the remark this answers is not in the document]
Comment (ggomez-cambriasolutions) [reply app/javascript/Application/styles/global/colors.sass]: you know, it's like black, but lighter. :)

File: app/javascript/Application/components/common/AssessmentRecordStatus.test.js [hunk 18]
```diff
@@ -10,9 +10,9 @@ describe('AssessmentRecordStatus', () => {
         status: AssessmentStatus.inProgress,
       })
       expect(expectedResult).toEqual(
-        <div className="status-icon-wrapper">
-          <Icon name="circle-notch" set={'fa'} className={`fa-2x`} />
-          <span className="assessment-in-progress">In Progress</span>
+        <div className={'status-icon-wrapper'}>
```
[the remark this answers is not in the document]
Comment (ggomez-cambriasolutions) [reply app/javascript/Application/components/common/AssessmentRecordStatus.test.js]: I didn't see any lint warnings for this component. What are your thoughts on this?

File: app/javascript/Application/components/common/Ellipsis.js [hunk 19]
```diff
@@ -23,31 +26,139 @@ export default class Ellipsis extends React.Component {
...
+
+    return isDeleteAssessmentWarningShown ? (
+      <PageModal
+        isOpen={true}
+        title={'Deleting CANS Warning'}
+        warningDescription={warning}
+        description={'This cannot be undone.'}
+        removeButtonLabel={'Delete CANS'}
+        cancelButtonLabel={'Cancel'}
+        onCancel={() => this.handleWarningCancel()}
```
[the remark this answers is not in the document]
Comment (ggomez-cambriasolutions) [reply app/javascript/Application/components/common/Ellipsis.js]: I redefined the methods that are passed as props within Ellipsis as anonymous functions. Is this what you meant, or is it always better to define the methods as anonymous functions?

File: app/javascript/Application/components/common/Ellipsis.js [hunk 20]
```diff
@@ -23,31 +26,139 @@ export default class Ellipsis extends React.Component {
...
+  async handleWarningDelete() {
+    const { assessmentId } = this.props
+    const { callback } = this.props
+    await AssessmentService.delete(assessmentId)
+    this.closeModal()
+    callback(assessmentId)
+  }
+
+  canDeleteAssessment() {
+    try {
```
[the remark this answers is not in the document]
Comment (ggomez-cambriasolutions) [reply app/javascript/Application/components/common/Ellipsis.js]: another case of being overly cautious. I refactored this piece of logic to use null checks, instead.

File: app/javascript/Application/components/common/Ellipsis.js [hunk 21]
```diff
@@ -23,31 +26,139 @@ export default class Ellipsis extends React.Component {
...
+
+  renderViewChangeLogMenuItem() {
+    return (
+      <div
+        className={'popover-view-change-log'}
+        onClick={e => this.onItemSelect(e)}
+        onKeyPress={e => this.handleKeyPress(e)}
+        role={'menuitem'}
+        aria-label="View Change Log Menu Item"
+        tabIndex={0}
```
[the remark this answers is not in the document]
Comment (ggomez-cambriasolutions) [reply app/javascript/Application/components/common/Ellipsis.js]: Yes, you are correct. It makes more sense to use a button element for each menu item. Done. :)

File: app/javascript/Application/components/Client/AssessmentHistory/ClientAssessmentHistoryTableEllipsis.js (hunk 15 above)
Comment (ggomez-cambriasolutions) [reply app/javascript/Application/components/Client/AssessmentHistory/ClientAssessmentHistoryTableEllipsis.js]: Good point. If the Ellipsis component is being rendered, then that means there is an assessment  associated with it. So, there really isn't a valid situation that these props would be falsy. Otherwise we wouldn't be rendering an Ellipsis component in the first place.

File: app/javascript/Application/components/common/Ellipsis.test.js (hunk 2 above)
Comment (ggomez-cambriasolutions) [reply app/javascript/Application/components/common/Ellipsis.test.js]: you're right, @pokharelmanish . popover is one word. :)

File: app/javascript/Application/components/common/Ellipsis.js (hunk 1 above)
Comment (ggomez-cambriasolutions) [reply app/javascript/Application/components/common/Ellipsis.js]: great suggestion, @pokharelmanish

File: app/javascript/Application/components/common/Ellipsis.js (hunk 20 above)
[the remark this answers is not in the document]
Comment (wmitchell-cambria) [reply app/javascript/Application/components/common/Ellipsis.js]: While I think of it, I should give two actual reasons why the null checks are better than the try/catch:

1. Performance can be slow on certain engines. For example, V8 used to skip optimizing functions that had try/catch. Chrome handles this well now, but that is only one of the ~5 engines we need to support.
2. Linear flow. A try/catch represents a possibility that any command will jump you down to the catch. Handling specific errors with null checks lets you read from left to right, top to bottom. That plays into complexity metrics for Code Climate, for example.

File: app/javascript/Application/components/Assessment/ChangeLog/ChangeLogLoadingBoundary.js (hunk 4 above)
Comment (wmitchell-cambria) [reply app/javascript/Application/components/Assessment/ChangeLog/ChangeLogLoadingBoundary.js]: The loading boundary changes look good!
